core/src/main/java/org/infinispan/util/concurrent/BlockingTaskAwareExecutorServiceImpl.java

```diff
@@ -142,4 +116,79 @@ private void doExecute(BlockingRunnable runnable) {
          blockedTasks.offer(runnable);
       }
    }
+
+   private class ControllerThread extends Thread {
+
+      private final Semaphore semaphore;
+      private volatile boolean interrupted;
+
+      public ControllerThread(String controllerThreadName) {
+         super(controllerThreadName);
+         semaphore = new Semaphore(0);
+      }
+
+      public void checkForReadyTask() {
+         semaphore.release();
+      }
+
+      @Override
+      public void interrupt() {
+         interrupted = true;
+         super.interrupt();
+         semaphore.release();
+
+      }
+
+      @Override
+      public void run() {
+         while (!interrupted) {
+            try {
+               semaphore.acquire();
+            } catch (InterruptedException e) {
+               return;
+            }
+            semaphore.drainPermits();
+            int size = blockedTasks.size();
+            if (size == 0) {
+               continue;
+            }
+            ArrayDeque<BlockingRunnable> readyList = new ArrayDeque<>(size);
+            for (Iterator<BlockingRunnable> iterator = blockedTasks.iterator(); iterator.hasNext(); ) {
+               BlockingRunnable runnable = iterator.next();
+               if (runnable.isReady()) {
+                  iterator.remove();
+                  readyList.addLast(runnable);
+               }
+            }
+
+            if (log.isTraceEnabled()) {
+               log.tracef("Tasks to be executed=%s, still pending=~%s", readyList.size(), size);
+            }
+
+            BlockingRunnable runnable;
+            while ((runnable = readyList.pollFirst()) != null) {
+               doExecute(runnable);
```

Comment:
Please change the default to AbortPolicy then, CallerRunsPolicy is ok if we overflow to the OOB thread pool, but not if it blocks all other commands.